srml/elections-phragmen/src/lib.rs

```diff
@@ -0,0 +1,952 @@
+			ensure_root(origin)?;
+			let who = T::Lookup::lookup(who)?;
+
+			if Self::is_candidate(&who) {
+				let members = Self::members();
+				let new_members: Vec<T::AccountId> = members
+					.into_iter()
+					.filter(|i| *i != who)
+					.collect();
+				<Members<T>>::put(new_members.clone());
+				T::Currency::unreserve(&who, T::CandidacyBond::get());
+				T::ChangeMembers::change_members(&[], &[who], new_members);
+			}
+		}
+
+		/// Set the presentation duration. If there is current a vote being presented for, will
+		/// invoke `finalize_vote`.
+		#[weight = SimpleDispatchInfo::FixedOperational(10_000)]
+		fn set_term_duration(origin, #[compact] count: T::BlockNumber) {
+			ensure_root(origin)?;
+			<TermDuration<T>>::put(count);
+		}
+
+		fn on_initialize(n: T::BlockNumber) {
+			if let Err(e) = Self::end_block(n) {
+				runtime_io::print("Guru meditation");
+				runtime_io::print(e);
+			}
+		}
+	}
+}
+
+decl_event!(
+	pub enum Event<T> where <T as system::Trait>::AccountId {
+		/// A new term with new members.
+		NewTerm(Vec<AccountId>),
+		/// No candidates were elected for this round.
+		EmptyCouncil(),
+	}
+);
+
+impl<T: Trait> Module<T> {
+	/// Check if `who` is a candidate.
+	fn is_candidate(who: &T::AccountId) -> bool {
+		Self::candidates().iter().find(|c| *c == who).is_some()
+	}
+
+
+	/// Check if `who` is a voter in the ongoing round.
+	fn is_current_voter(who: &T::AccountId) -> bool {
+		let now = Self::election_rounds();
+		Self::votes_of(who).1 == now && <StakeOf<T>>::exists(who)
+	}
+
+	/// Check if `who` was a voter who has not yet called `remove_voter`.
+	fn is_old_voter(who: &T::AccountId) -> bool {
+		let now = Self::election_rounds();
+		Self::votes_of(who).1 < now && <StakeOf<T>>::exists(who)
+	}
+
+	/// The locked stake of a voter.
+	fn locked_stake_of(who: &T::AccountId) -> BalanceOf<T> {
+		if Self::is_current_voter(who) {
+			Self::stake_of(who)
+		} else {
+			Zero::zero()
+		}
+	}
+
+	/// Check there's nothing to do this block.
+	///
+	/// Runs phragmen election and cleans all the previous candidate state. The voter state is NOT
+	/// cleaned and voters must themselves submit a transaction to clean it.
+	fn end_block(block_number: T::BlockNumber) -> Result {
+		if (block_number % Self::term_duration()).is_zero() {
+			use sr_primitives::phragmen;
+			let seats = Self::desired_seats() as usize;
+			let candidates = Self::candidates();
+			let current_round = Self::election_rounds();
+			let voters_and_votes = <VotesOf<T>>::enumerate()
+				.filter_map(|(v, i)| if i.1 == current_round { Some((v, i.0)) } else { None } )
+				.collect::<Vec<(T::AccountId, Vec<T::AccountId>)>>();
+
+			let maybe_new_members = phragmen::elect::<_, _, _, T::CurrencyToVote>(
+				seats,
+				false,
+				0,
+				candidates.clone(),
+				voters_and_votes,
+				Self::locked_stake_of
+			);
+
+			let old_members = <Members<T>>::take();
+			if let Some((new_members_with_approval, _)) = maybe_new_members {
```

Comment:
we should probably sort both of these to turn it from O(N**2) to O(NlogN).

Reply:
great point -- this + `is_candidate()` now operates on sorted lists, hence being _O(NlogN)_, hence making the `submit_candidate()` also _O(NlogN)_.